jsonparse.js

```diff
@@ -261,17 +266,19 @@ proto.write = function (buffer) {
             break;
           default:
             this.tState = START;
-            var result = Number(this.string);
-
-            if (isNaN(result)){
+            if (isNaN(this.string)){
               return this.charError(buffer, i);
             }
 
-            if ((this.string.match(/[0-9]+/) == this.string) && (result.toString() != this.string)) {
-              // Long string of digits which is an ID string and not valid and/or safe JavaScript integer Number
-              this.onToken(STRING, this.string);
+            if (this.parseNumbersAsStrings) {
+              this.onToken(STRING, this.string)
             } else {
-              this.onToken(NUMBER, result);
+              const result = Number(this.string)
+              if (precisionWasLost(this.string, result)) {
+                this.onError(new Error( `${this.string} is unsafe to parse as a number because it is either too large (positive or negative), has too many significant figures, or is not finite when parsed. Please pass in the {parseNumbersAsStrings: true} option`))
```

Comment:
```suggestion
                this.onError(new Error(`${this.string} is unsafe to parse as a number because it is either too large (positive or negative), has too many significant figures, or is not finite when parsed. Please pass in the {parseNumbersAsStrings: true} option`))
```